File: backend/08-review/src/main/kotlin/com/dokandar/review/observability/Log.kt
package com.dokandar.review.observability

import com.dokandar.review.Config
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import com.mongodb.client.MongoCollection
import org.bson.Document
import co.elastic.apm.api.ElasticApm
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Base64
import java.util.concurrent.ConcurrentLinkedQueue
import kotlin.concurrent.thread

// Three-sink structured logging mirroring 01-auth/04-catalog: stdout JSON w/ elasticapm_* fields,
// MongoDB forensic collection <service>, Elasticsearch data stream logs-app-08-review-* (:9200).
// Trace ids read from SLF4J MDC — the Elastic APM -javaagent injects trace.id/transaction.id/span.id.
object Log {
    private val mongoQ = ConcurrentLinkedQueue<Document>()
    private val esQ = ConcurrentLinkedQueue<Map<String, Any?>>()
    @Volatile private var mongoColl: MongoCollection<Document>? = null
    @Volatile private var mongoUp = false
    private val http: HttpClient = HttpClient.newHttpClient()
    private val ts = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss,SSS")
    private val access = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss")

    val mongoHealthy: Boolean get() = mongoUp

    private fun apmFields(): Map<String, Any?> {
        var trace: String? = null; var tx: String? = null; var span: String? = null
        try {
            val s = ElasticApm.currentSpan(); trace = s.traceId.ifEmpty { null }; span = s.id.ifEmpty { null }
            tx = ElasticApm.currentTransaction().id.ifEmpty { null }
        } catch (_: Throwable) {}
        val m = linkedMapOf<String, Any?>(
            "elasticapm_service_name" to Config.apmServiceName,
            "elasticapm_service_environment" to Config.appEnv,
        )
        if (trace != null) m["elasticapm_trace_id"] = trace
        if (tx != null) m["elasticapm_transaction_id"] = tx
        m["elasticapm_labels"] = linkedMapOf(
            "transaction.id" to tx, "trace.id" to trace, "span.id" to span,
            "service.name" to Config.apmServiceName, "service.environment" to Config.appEnv,
        )
        return m
    }

    fun info(name: String, msg: String) = emit("INFO", name, msg)
    fun warn(name: String, msg: String) = emit("WARNING", name, msg)
    fun error(name: String, msg: String) = emit("ERROR", name, msg)

    private fun emit(level: String, name: String, msg: String) {
        val now = ZonedDateTime.now()
        val af = apmFields()
        val stdout = linkedMapOf<String, Any?>("asctime" to now.format(ts), "name" to name, "levelname" to level, "message" to msg)
        stdout.putAll(af)
        println(Json.encode(stdout))
        val trace = af["elasticapm_trace_id"]
        val doc = linkedMapOf<String, Any?>(
            "@timestamp" to now.withZoneSameInstant(java.time.ZoneOffset.UTC).format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")),
            "log" to mapOf("level" to level.lowercase(), "logger" to name),
            "message" to msg,
            "service" to mapOf("name" to Config.serviceName, "version" to Config.codeVersion, "environment" to Config.appEnv),
            "labels" to mapOf("tenant" to Config.tenant, "env_version" to Config.envVersion),
        )
        doc.putAll(af)
        if (trace != null) { doc["trace"] = mapOf("id" to trace); doc["transaction"] = mapOf("id" to af["elasticapm_transaction_id"]) }
        if (Config.mongoLogUri.isNotEmpty() && mongoQ.size < 5000) mongoQ.add(Document(doc))
        if (Config.esUrl.isNotEmpty() && esQ.size < 5000) esQ.add(doc)
    }

    fun access(ip: String, method: String, path: String, status: Int, reason: String) =
        println("${ZonedDateTime.now().format(access)}    $ip - \"$method $path HTTP/1.1\" $status $reason")

    fun startSinks() {
        if (Config.mongoLogUri.isNotEmpty()) {
            try {
                val c: MongoClient = MongoClients.create(Config.mongoLogUri)
                mongoColl = c.getDatabase(Config.mongoLogDb).getCollection(Config.serviceName); mongoUp = true
            } catch (e: Exception) { System.err.println("mongo log sink connect failed: ${e.message}") }
            thread(isDaemon = true, name = "mongo-log-sink") {
                while (true) {
                    Thread.sleep(2000)
                    val coll = mongoColl ?: continue
                    val batch = ArrayList<Document>(); while (batch.size < 500) { val d = mongoQ.poll() ?: break; batch.add(d) }
                    if (batch.isEmpty()) continue
                    try { coll.insertMany(batch); mongoUp = true } catch (e: Exception) { mongoUp = false }
                }
            }
        }
        if (Config.esUrl.isNotEmpty()) {
            val ds = "logs-app-${Config.serviceName}-default"
            val url = "${Config.esUrl.trimEnd('/')}/$ds/_bulk"
            val auth = if (Config.esUser.isNotEmpty()) "Basic " + Base64.getEncoder().encodeToString("${Config.esUser}:${Config.esPassword}".toByteArray()) else null
            thread(isDaemon = true, name = "es-log-sink") {
                while (true) {
                    Thread.sleep(2000)
                    val batch = ArrayList<Map<String, Any?>>(); while (batch.size < 500) { val d = esQ.poll() ?: break; batch.add(d) }
                    if (batch.isEmpty()) continue
                    val sb = StringBuilder()
                    for (d in batch) { sb.append("{\"create\":{}}\n"); sb.append(Json.encode(d)); sb.append("\n") }
                    try {
                        val b = HttpRequest.newBuilder(URI.create(url)).header("content-type", "application/x-ndjson")
                        if (auth != null) b.header("Authorization", auth)
                        http.send(b.POST(HttpRequest.BodyPublishers.ofString(sb.toString())).build(), HttpResponse.BodyHandlers.ofString())
                    } catch (_: Exception) {}
                }
            }
        }
    }
}

File: backend/09-payment/lib/payment/obs/log.ex
defmodule Payment.Obs.Log do
  @moduledoc "3-sink ECS logger (stdout + Mongo + ES :9200) with OTel trace correlation."
  use GenServer
  require Record
  alias Payment.Config

  Record.defrecordp(:span_ctx, Record.extract(:span_ctx, from_lib: "opentelemetry_api/include/opentelemetry.hrl"))

  def start_link(_), do: GenServer.start_link(__MODULE__, %{mongo_q: [], es_q: [], mongo_up: false}, name: __MODULE__)

  def info(name, msg), do: emit("INFO", name, msg)
  def warn(name, msg), do: emit("WARNING", name, msg)
  def error(name, msg), do: emit("ERROR", name, msg)
  def mongo_healthy?, do: (try do GenServer.call(__MODULE__, :mongo_up, 500) rescue _ -> false catch _, _ -> false end)
  def access(ip, method, path, status, reason) do
    ts = Calendar.strftime(DateTime.utc_now(), "%d-%m-%Y %H:%M:%S")
    IO.puts("#{ts}    #{ip} - \"#{method} #{path} HTTP/1.1\" #{status} #{reason}")
  end

  defp trace_ids do
    try do
      case :otel_tracer.current_span_ctx() do
        ctx when Record.is_record(ctx, :span_ctx) ->
          tid = span_ctx(ctx, :trace_id); sid = span_ctx(ctx, :span_id)
          if tid in [0, :undefined] or is_nil(tid), do: {nil, nil}, else: {hex(tid, 32), hex(sid, 16)}
        _ -> {nil, nil}
      end
    rescue _ -> {nil, nil} catch _, _ -> {nil, nil} end
  end
  defp hex(int, w), do: :io_lib.format("~#{w}.16.0b", [int]) |> List.to_string()

  defp emit(level, name, msg) do
    {trace, span} = trace_ids()
    now = DateTime.utc_now()
    apm = if trace, do: %{
      "elasticapm_trace_id" => trace, "elasticapm_transaction_id" => trace, "elasticapm_span_id" => span,
      "elasticapm_service_name" => Config.apm_service_name(), "elasticapm_service_environment" => Config.app_env(),
      "elasticapm_labels" => %{"trace.id" => trace, "transaction.id" => trace, "span.id" => span, "service.name" => Config.apm_service_name(), "service.environment" => Config.app_env()}
    }, else: %{"elasticapm_service_name" => Config.apm_service_name(), "elasticapm_service_environment" => Config.app_env(), "elasticapm_labels" => %{"trace.id" => nil, "transaction.id" => nil, "span.id" => nil, "service.name" => Config.apm_service_name(), "service.environment" => Config.app_env()}}
    stdout = Map.merge(%{"asctime" => Calendar.strftime(now, "%Y-%m-%d %H:%M:%S,") <> (now.microsecond |> elem(0) |> div(1000) |> Integer.to_string() |> String.pad_leading(3, "0")), "name" => name, "levelname" => level, "message" => msg}, apm)
    IO.puts(Jason.encode!(stdout))
    doc = %{"@timestamp" => DateTime.to_iso8601(now), "log" => %{"level" => String.downcase(level), "logger" => name}, "message" => msg,
            "service" => %{"name" => Config.service_name(), "version" => Config.code_version(), "environment" => Config.app_env()},
            "labels" => %{"tenant" => Config.tenant(), "env_version" => Config.env_version()}}
    doc = if trace, do: Map.merge(doc, %{"trace" => %{"id" => trace}, "transaction" => %{"id" => trace}, "elasticapm_trace_id" => trace, "elasticapm_transaction_id" => trace, "elasticapm_service_name" => Config.apm_service_name(), "elasticapm_service_environment" => Config.app_env()}), else: doc
    (try do GenServer.cast(__MODULE__, {:ship, doc}) rescue _ -> :ok catch _, _ -> :ok end)
  end

  @impl true
  def init(state) do
    if Config.mongo_log_uri() != "", do: send(self(), :connect_mongo)
    schedule()
    {:ok, state}
  end
  @impl true
  def handle_call(:mongo_up, _from, s), do: {:reply, s.mongo_up, s}
  @impl true
  def handle_cast({:ship, doc}, s) do
    s = if Config.mongo_log_uri() != "" and length(s.mongo_q) < 5000, do: %{s | mongo_q: [doc | s.mongo_q]}, else: s
    s = if Config.es_url() != "" and length(s.es_q) < 5000, do: %{s | es_q: [doc | s.es_q]}, else: s
    {:noreply, s}
  end
  @impl true
  def handle_info(:connect_mongo, s) do
    case (try do Mongo.start_link(url: mongo_url()) rescue _ -> :error catch _, _ -> :error end) do
      {:ok, pid} -> {:noreply, %{s | mongo_up: true} |> Map.put(:mongo_pid, pid)}
      _ -> Process.send_after(self(), :connect_mongo, 5000); {:noreply, s}
    end
  end
  defp mongo_url do
    uri = Config.mongo_log_uri(); db = Config.mongo_log_db()
    cond do
      String.contains?(uri, "/?") -> String.replace(uri, "/?", "/#{db}?")
      String.ends_with?(uri, "/") -> uri <> db
      true -> uri <> "/" <> db
    end
  end
  def handle_info(:flush, s) do
    s = flush_mongo(s); s = flush_es(s); schedule(); {:noreply, s}
  end
  defp schedule, do: Process.send_after(self(), :flush, 2000)

  defp flush_mongo(%{mongo_pid: pid} = s) when is_pid(pid) do
    {batch, rest} = Enum.split(Enum.reverse(s.mongo_q), 500)
    if batch != [] do
      up = (try do Mongo.insert_many(pid, Config.service_name(), batch); true rescue _ -> false catch _, _ -> false end)
      %{s | mongo_q: Enum.reverse(rest), mongo_up: up}
    else s end
  end
  defp flush_mongo(s), do: s

  defp flush_es(s) do
    if Config.es_url() == "" do s else
      {batch, rest} = Enum.split(Enum.reverse(s.es_q), 500)
      if batch != [] do
        body = Enum.map_join(batch, "", fn d -> ~s({"create":{}}\n) <> Jason.encode!(d) <> "\n" end)
        url = String.trim_trailing(Config.es_url(), "/") <> "/logs-app-#{Config.service_name()}-default/_bulk"
        auth = if Config.es_user() != "", do: [auth: {:basic, "#{Config.es_user()}:#{Config.es_password()}"}], else: []
        (try do Req.post(url, [headers: [{"content-type", "application/x-ndjson"}], body: body, retry: false] ++ auth) rescue _ -> :ok catch _, _ -> :ok end)
        %{s | es_q: Enum.reverse(rest)}
      else s end
    end
  end
end
